File: src/Java/1-100/69.MySqrt.java
import java.awt.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import javax.lang.model.util.ElementScanner6;

class Solution {
    public int mySqrt(int x) {
        if (x == 0)
            return 0;
        if (x == 1)
            return 1;
        int left = 1;
        int right = x / 2;
        while (left < right) {
            int middle = left + ((right - left + 1) >> 1);
            if (middle == x / middle)
                return middle;
            else if (middle < x / middle)
                left = middle;
            else
                right = middle - 1;
        }
        return left;
    }
}

File: src/Python/1-100/69.MySqrt.py

class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x == 0:
            return 0
        if x == 1:
            return 1
        l = 1
        r = x // 2
        # 二分查找
        while l < r:
            mid = l + (( r - l + 1) >> 1)
            if mid * mid == x: 
                return mid
            elif mid * mid < x:
                l = mid
            else:
                r = mid - 1
        return l

if __name__ == '__main__':
    solution = Solution()
    print(solution.mySqrt(4))
    print(solution.mySqrt(8))
else:
    pass
